### ISSUE TITLE
Negative discrepancy

### ISSUE THREAD

**Issue by asardaes** [mblondel/soft-dtw#5]
Hello, I'm testing this python version of soft-DTW and I am seeing the following:

```python
import sdtw
import numpy as np
x = np.array([0,0,0,-0.033203,-0.057491,-0.08566,-0.11358,-0.13903,-0.16222,-0.18402])
y = np.array([0,0,0,-0.048463,-0.073921,-0.10089,-0.12832,-0.1578,-0.19184,-0.23162])
x = x.reshape(-1,1)
y = y.reshape(-1,1)
from sklearn.metrics.pairwise import euclidean_distances
D = euclidean_distances(x,y, squared=True)
sdtw_obj = sdtw.SoftDTW(D)
sdtw_obj.compute()
#> -14.175522634015779
```

It seems counter intuitive to obtain negative values for a discrepancy, but maybe it is normal in this case? I would like to corroborate.

**Comment by mblondel** [comment]
This is expected behavior. The smoothing of the min based on the log-sum-exp can lead to negative values. In addition, like the original DTW, soft-DTW doesn't fulfill the triangle inequality.

**Comment by asardaes** [comment]
Ok, thanks for the confirmation.

On a similar note, if I set `y` to `x.copy()` I get a non-zero value. Is this also expected?

**Comment by mblondel** [comment]
Indeed, soft-DTW(x, x) is not zero because soft-DTW considers all paths, not just the diagonal one. soft-DTW is most useful when you want to differentiate it.

**Comment by asardaes** [comment]
I see, thanks for the clarification. One last question if you don't mind: is soft-DTW always symmetric?

**Comment by mblondel** [comment]
It is.

**Comment by asardaes** [comment]
Thanks!

**Comment by Maghoumi** [comment]
I know this issue was closed, but could you @mblondel kindly clarify something please:  
When used as a loss metric, how should one interpret this negative dissimilarity number?

Imagine that my goal is to reduce the soft-DTW dissimilarity between two sequences `x` and `y` using gradient descent. If the dissimilarity score for some values of `x` and `y` is negative, how should one proceed? Can I always minimize the `abs(sdtw(x, y))` or do I need to handle negative cases as a special case?

Thanks in advance,

**Comment by mblondel** [comment]
Essentially, soft-dtw returns as value the log partition function of a probabilistic model (more precisely a conditional random field over alignments). This is why the value can be negative, just as the log likelihood can be negative.

If you care about non-negativity, a debiasing/normalization is proposed by @marcocuturi in #10.

**Comment by Maghoumi** [comment]
Thanks a lot for the clarification and the prompt response @mblondel.   

I had seen @marcocuturi's formula and the blog post, but I think I need to study it more to understand what it does and how it's applicable to my problem.

**Comment by stellarpower** [comment]
Is a negative loss still usable? I.e. that as the loss continues to drop more negative, the signals are closer together than they were? Or is it something that needs correcting for in gradient descent?
